src/nanoarrow/nanoarrow.hpp

```diff
@@ -273,6 +280,49 @@ using UniqueArrayView = internal::Unique<struct ArrowArrayView>;
 
 /// @}
 
+/// \defgroup nanoarrow_hpp-buffer Buffer helpers
+///
+/// Helpers to wrap buffer-like C++ objects as ArrowBuffer objects that can
+/// be used to build ArrowArray objects.
+///
+/// @{
+
+/// \brief Initialize a buffer wrapping an arbitrary C++ object
+///
+/// Initializes a buffer with a release callback that deletes the moved obj
+/// when ArrowBufferReset is called. T must be movable.
+template <typename T>
+static inline void BufferInitWrapped(struct ArrowBuffer* buffer, T obj, const void* ptr,
+                                     int64_t size_bytes) {
+  T* obj_moved = new T(std::move(obj));
+  buffer->data = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(ptr));
+  buffer->size_bytes = size_bytes;
+  buffer->capacity_bytes = 0;
+  buffer->allocator =
+      ArrowBufferDeallocator(&internal::DeallocateWrappedBuffer<T>, obj_moved);
+}
```

Comment:
what kind of object is intended to be passed in here that would be valid to use as the data pointer?

Reply:
the problem with that is that it wouldn't be valid for that to be the `buffer->data` in the resulting `ArrowArray` object. Shouldn't that be placed in the `private_data`, not the `data`?